src/yarn.ts

```diff
@@ -8,99 +8,192 @@
+     * @param children the list of dependencies to analyze
+     * @param nodeTreeDependencies the object containing the tree of dependencies
+     * @param subsetDependencies  the
      */
-    protected async addNodePackage(nodeModulesFolder: string, yarnNode: IYarnNode, packages: INodePackage[]): Promise<void> {
+    protected findDependencies(children: string[], nodeTreeDependencies: Map<string, string[]>, subsetDependencies: string[]): void {
+
+        children.map(child => {
+            // only loop on exist
+            if (subsetDependencies.indexOf(child) >= 0) {
+                return;
+            }
+            subsetDependencies.push(child);
+
+            // loop on children in any
+            let depChildren = nodeTreeDependencies.get(child);
+            if (depChildren) {
+                depChildren = depChildren.filter(depChild => {
+                    const res = this.excludedPackages.indexOf(depChild) < 0;
+                    if (!res) {
+                        Logger.debug(` --> Excluding the dependency ${depChild}`);
+                    }
+                    return res;
+                });
+
+                const matching: string[] = [];
+                const foundForbiddenPackage = depChildren.some(r => {
+                    const res = this.forbiddenPackages.indexOf(r) >= 0;
+                    if (res) {
+                        matching.push(r);
+                    }
+                    return res;
+                });
+                if (foundForbiddenPackage) {
+                    throw new CliError(`Forbidden dependencies ${matching} has been found as dependencies of ${child}` +
+                        `Current dependencies: ${depChildren}, excluded list: ${this.forbiddenPackages}`);
+                }
+                this.findDependencies(depChildren, nodeTreeDependencies, subsetDependencies);
+            }
+        }
+        );
+    }
 
-        // add each child to the array again
+    /**
+     * Insert the given node into the Map/tree of dependencies
+     * @param yarnNode the node to insert
+     * @param nodeTreeDependencies the tree to update
+     */
+    protected insertNode(yarnNode: IYarnNode, nodeTreeDependencies: Map<string, string[]>): void {
+        const npmModuleName = yarnNode.name.substring(0, yarnNode.name.lastIndexOf('@'));
+
+        // init dependencies object if not existing
+        let dependencies = nodeTreeDependencies.get(npmModuleName);
+        if (!dependencies) {
+            dependencies = [];
+            nodeTreeDependencies.set(npmModuleName, dependencies);
+        }
+
+        // insert all children as well
         if (yarnNode.children) {
-            yarnNode.children.forEach((child) => {
-                this.addNodePackage(nodeModulesFolder, child, packages);
+            yarnNode.children.map(child => {
+                const childName = child.name.substring(0, child.name.lastIndexOf('@'));
```

Comment:
Only little question: so, we grab dependencies and version doesn't matter?

Reply:
yes. It's because dependencies not shared/hosted will be in node_modules of dependencies so they'll be added automatically as well
This is what is done for `che:theia production`